Change kubecontext section to improve performance

reegnz opened the pull request:
#### Description

The kubecontext section script called kubectl config twice, and it's not a
cheap call to make.

On my machine the call takes roughly 300ms, two consecutive calls make for a
pretty heavy 600 ms then.

Now I have reduced the section to only call kubectl config only once, making
the section spend roughly half the time to calculate the kubecontext display
text.

reegnz commented:
Nobody is troubled by the slowness of the kube-prompt?
Or do I need to change something / sign something / etc. to get this merged?

reegnz commented:
I checked jonmosco/kube-ps1 and it seems that they are caching the context in a smarter way:
* check current context with kubectl
* cache check time, context + namespace
* further checks are short-circuited if config file didn't change.

It relies on kubectl implementation detail though, but we could safely assume it is either $HOME/.kube/config or files defined with $KUBECONFIG.

reegnz commented:
Incorporated the caching idea from kube-ps1 into the change.
Now the speed of the kubecontext section is close to instantaneous (exept right after switching kube context).

reegnz commented:
@denysdovhan please check this out! I am not exaggerating when I say it's a **huge** improvement over the existing solution.

reegnz commented:
Seems like the multiple-kubecontext file use-case still needs some adjustments. WIP

denysdovhan commented:
@reegnz please, let me know when this PR will be ready. I'd be grateful if someone who uses kubecontext would check this PR. I personally don't use this tool.

maximbaz commented:
I use this section, to be honest I'm not super fan of this change polluting ~/.kube folder, but I realize that I probably don't feel your pain because I use the async version of spaceship :)

maybe at the very least let's make this respect XDG_CACHE_HOME?

reegnz commented:
@maximbaz yeah, xdg-cache makes more sense.
kube-ps1 uses  variables to store the cached state though, maybe that's also an option?

reegnz commented:
Done, XDG_CACHE_HOME is used per [specification](https://specifications.freedesktop.org/basedir-spec/0.6/ar01s03.html).

As for making it fileless: I have no idea how to set a global variable from spaceship-prompt, whenever I try to set a variable it resets on each invocation.

reegnz commented:
@denysdovhan I think I finished everything.

how to test:
* Have kubectl and [kubectx/kubens](https://github.com/ahmetb/kubectx) installed.
* good to have multiple kubernetes contexts at hand

1. When switching to context, it should be displayed.
```sh
➜ kubectx minikube
Switched to context "minikube".

~ at ☸️  minikube
➜
```

2. When switching namespaces it should display next to the context (except for `default`, that is hidden).
```sh
~ at ☸️  minikube
➜ kubens kube-system
Context "minikube" modified.
Active namespace is "kube-system".

~ at ☸️  minikube (kube-system) took 2s
➜ kubens default
Context "minikube" modified.
Active namespace is "default".

~ at ☸️  minikube
➜
```

3. When switching contexts, the new context should show up.
```
~ at ☸️  minikube
➜ kubectx EKS-dev
Switched to context "EKS-dev".

~ at ☸️  EKS-dev
➜
```
4. If the context is named default, it shouldn't be hidden (only default namespace name is hidden).
```
~ at ☸️  minikube
➜ export KUBECONFIG="$HOME/.kube/config:$(k3d get-kubeconfig --name='k3s-default')"

~ at ☸️  minikube
➜ kubectx default
Switched to context "default".

~ at ☸️  default
```

maximbaz commented:
The code looks good, ok with merging if @denysdovhan likes this 👍 

Personally I'm not a fan of letting a section implement its own caching mechanism, I think the work on making spaceship should be prioritized instead (which would solve this issue), and if you would still want to have the cache afterwards, it better be implemented by spaceship core so that it's available to all sections. But that's just my 5 cents 🙂

reegnz commented:
I believe caching is most of the time tool dependent. I find it hard to see how you could generalize this.
For this use-case it was trivial and kube-ps1 was a shining example on how to do it. For something like docker it would be harder to cache.

Also regarding async, 200 ms for  fetching a version is 200 ms even if it's done asynchronously. it just spares you from making the latency worse if you have multiple sections that are fetched in parallel. The slowest section will still determine how laggy your  prompt feels.
Even with this fix we have at least another section that is below 200ms on my machine (a pretty beefy 2018 MBP), and I do feel that this cannot be easily fixed on a global basis, only section by section.

solidnerd commented on sections/kubecontext.zsh:
```diff
@@ -18,24 +18,67 @@ SPACESHIP_KUBECONTEXT_SYMBOL="${SPACESHIP_KUBECONTEXT_SYMBOL="☸️  "}"
+}
+kubecontext_cache() {
+    [[ -d ${SPACESHIP_KUBECONTEXT_CACHE_DIR} ]] || mkdir -p ${SPACESHIP_KUBECONTEXT_CACHE_DIR}
+    [[ -r ${SPACESHIP_KUBECONTEXT_CACHE_FILE} ]] || touch -d @0 ${SPACESHIP_KUBECONTEXT_CACHE_FILE}
```
This doesn't work on os x.

```
touch: illegal option -- d
usage:
touch [-A [-][[hh]mm]SS] [-acfhm] [-r file] [-t [[CC]YY]MMDDhhmm[.SS]] file ...
stat: /Users/niclas/.cache/spaceship-prompt/kubecontext-cache: stat: No such file or directory
```

The problem here is that os x uses a different `touch`binary. I think it can be fixed with installing coreutils through brew and using `gtouch` instead of `touch`. So a command checker is needed if touch or gtouch should be used.

reegnz replied:
That's embarassing, since I'm on os x but use gnu utils on my path. :)
I will see to how I can fix that without  gnu touch.

I guess `touch -t 197001011200` would also work.

reegnz replied:
OK, I will change it to:
`TZ=UTC touch -t 197001010000 ${SPACESHIP_KUBECONTEXT_CACHE_FILE}`, that seems equivalent.

j16sdiz commented:
This PR now conflict with head

reegnz commented:
@j16sdiz Feel tree to pick it up if still relevant. Otherwise I'm ok with a PR reject.
I've since moved on to [p10k](https://github.com/romkatv/powerlevel10k).

dagadbm commented:
why doesn't this just read the files on .kube ? the kubectl is unbearably slow lol. impossible to use

reegnz commented:
@dagadbm and how would you exactly parse the kube config without introducing dependencies? You'd need a yaml parser for that. I wouldn't write a yaml parser in shell script.

dagadbm commented:
I thought a simple regex would do it but maybe not.. But are all shells
with kube info slow?

On Wed, Mar 4, 2020, 16:23 Zoltán Reegn <notifications@github.com> wrote:

> @dagadbm <https://github.com/dagadbm> and how would you exactly parse the
> kube config without introducing dependencies? You'd need a yaml parser for
> that. I wouldn't write a yaml parser in shell script.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/denysdovhan/spaceship-prompt/pull/690?email_source=notifications&email_token=AD7WGR7WB24FUBZNUUOTTODRFZ6ARA5CNFSM4H3Q2ULKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOENYYGIY#issuecomment-594641699>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AD7WGRZZJUHTPAS2YQAZEDDRFZ6ARANCNFSM4H3Q2ULA>
> .
>

romkatv commented:
> I thought a simple regex would do it but maybe not.

Depends on what you define as simple.

> But are all shells with kube info slow?

Not, not all Zsh themes capable of displaying current kubernetes context are slow.
